#' Generates a bedfile from inFile
#'
#' Write a bedfile from Breakpoint.R files for upload on to UCSC Genome browser
#'
#' @param index A character used to name the bedfile(s).
#' @param outputDirectory Location to write bedfile(s).
#' @param fragments A \code{\link{GRanges-class}} object with strand and mapq metadata, such as that generated by \code{\link{readBamFileAsGRanges}}
#' @param deltaWs A \code{\link{GRanges-class}} object with metadata column "deltaW" generated by \code{\link{deltaWCalculator}}.
#' @param breakTrack A \code{\link{GRanges-class}} object with metadata "genoT" (e.g. newBreaks) will write a bedtrack with refined breakpoints.
#' @param confidenceIntervals A \code{\link{GRanges-class}} object with metadata "genoT" the same length as \code{breakTrack} (e.g. confint) will write a bedtrack with breakpoints confidence intervals.
#' @param maskedRegions A \code{\link{GRanges-class}} object.
#' @param hotspots A \code{\link{GRanges-class}} object.
#' @param breaksGraph A \code{\link{GRanges-class}} object.
#' @param col A string with two RGB values separated by a space specifying the color for the two strands.
#' @return \code{NULL}
#' @author Ashley Sanders, David Porubsky, Aaron Taudt
#' @importFrom utils write.table
#' @export
#' @examples
#'## Get an example file
#'exampleFolder <- system.file("extdata", "example_results", package="breakpointRdata")
#'exampleFile <- list.files(exampleFolder, full.names=TRUE)[1]
#'## Load the file 
#'brkpts <- get(load(exampleFile))
#'## Write results to BED files
#'writeBedFile(index='testfile', outputDirectory=tempdir(), breakTrack=brkpts$breaks)
         
writeBedFile <- function(index, outputDirectory, fragments=NULL, deltaWs=NULL, breakTrack=NULL, confidenceIntervals=NULL, maskedRegions=NULL, hotspots=NULL, breaksGraph=NULL, col="103,139,139 243,165,97") {
  
    ## Insert chromosome for in case it's missing
    insertchr <- function(gr) {
        mask <- which(!grepl('chr', seqnames(gr)))
        mcols(gr)$chromosome <- as.character(seqnames(gr))
        mcols(gr)$chromosome[mask] <- sub(pattern='^', replacement='chr', mcols(gr)$chromosome[mask])
        mcols(gr)$chromosome <- as.factor(mcols(gr)$chromosome)
        return(gr)
    }

    ## Write read fragments to file
    if (!is.null(fragments) & length(fragments) > 0) {
        fragments <- insertchr(fragments)
        savefile.reads <- file.path(outputDirectory, paste0(index, '_reads.bed.gz'))
        ptm <- startTimedMessage("Writing to file ", savefile.reads, " ...")
        savefile.reads.gz <- gzfile(savefile.reads, 'w')
        head_reads <- paste0('track name=', index, '_reads visibility=1 colorByStrand="', col, '"') 
        utils::write.table(head_reads, file=savefile.reads.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=FALSE, sep='\t')   
        if (length(fragments)>0) {
            bedfile <- as.data.frame(fragments)[c('chromosome','start','end', 'mapq', 'strand')]
            bedfile$name <- 0 # adds a column of 0 as the'name' in the bedfile
            bedfile <- bedfile[c('chromosome','start','end','name','mapq', 'strand')]
        } else {
            bedfile <- data.frame(chromosome='chr1', start=1, end=1, name=NA, mapq=0, strand='.')
        }
        utils::write.table(bedfile,file=savefile.reads.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=TRUE, sep='\t')
        close(savefile.reads.gz)
        stopTimedMessage(ptm)
    }

    ## Write breakPoints and confidence intervals to file
    if (!is.null(breakTrack) & length(breakTrack) > 0) {
        breakTrack <- insertchr(breakTrack)
        savefile.breakPoints <- file.path(outputDirectory, paste0(index, '_breakPoints.bed.gz'))
        ptm <- startTimedMessage("Writing to file ", savefile.breakPoints, " ...")
        savefile.breakPoints.gz <- gzfile(savefile.breakPoints, 'w')
        head_br <- paste('track name=', index, '_BPs description=BedGraph_of_breakpoints_',index,'_allChr visibility=dense color=75,125,180', sep="")
        utils::write.table(head_br, file=savefile.breakPoints.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=FALSE, sep='\t')
        if (length(breakTrack)>0) {
            bpG <-  as.data.frame(breakTrack)[c('chromosome','start','end','genoT')]
        } else {
            bpG <- data.frame(chromosome='chr1', start=1, end=1, genoT=NA)
        }
        if (!is.null(confidenceIntervals)) {
            bpG$score <- 0
            bpG$strand <- '.'
            bpG$thickStart <- bpG$start
            bpG$thickEnd <- bpG$end
            bpG$start <- start(confidenceIntervals)
            bpG$end <- end(confidenceIntervals)
        }
        utils::write.table(bpG, file=savefile.breakPoints.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=TRUE, sep='\t')
        close(savefile.breakPoints.gz)
        stopTimedMessage(ptm)
    }

    ## Write deltaWs to file
    if (!is.null(deltaWs) & length(deltaWs) > 0) {
        deltaWs <- insertchr(deltaWs)
        savefile.deltaWs <- file.path(outputDirectory, paste0(index, '_deltaWs.bed.gz'))
        ptm <- startTimedMessage("Writing to file ", savefile.deltaWs, " ...")
        savefile.deltaWs.gz <- gzfile(savefile.deltaWs, 'w')
        head_dW <- paste('track type=bedGraph name=', index,'_dWs description=BedGraph_of_deltaWs_',index, '_allChr visibility=full color=200,100,10', sep="")
        utils::write.table(head_dW, file=savefile.deltaWs.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=FALSE, sep='\t')   
        if (length(deltaWs)>0) {
            bedG <- as.data.frame(deltaWs)[c('chromosome','start','end','deltaW')]
        } else {
            bedG <- data.frame(chromosome='chr1', start=1, end=1, deltaW=NA)
        }
        utils::write.table(bedG, file=savefile.deltaWs.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=TRUE, sep='\t')
        close(savefile.deltaWs.gz)
        stopTimedMessage(ptm)
    }

    ## Write maskedRegions to file
    if (!is.null(maskedRegions) & length(maskedRegions) > 0) {
        maskedRegions <- insertchr(maskedRegions)
        savefile.masked <- file.path(outputDirectory, paste0(index, '_masked.bed.gz'))
        ptm <- startTimedMessage("Writing to file ", savefile.masked, " ...")
        savefile.masked.gz <- gzfile(savefile.masked, 'w')
        head_masked <- paste('track name=', index,' description=BedGraph_of_maskedRegions_',index, '_allChr visibility=1 color=0,0,0', sep="")
        utils::write.table(head_masked, file=savefile.masked.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=FALSE, sep='\t')   
        if (length(maskedRegions)>0) {
            bedG <- as.data.frame(maskedRegions)[c('chromosome','start','end')]
        } else {
            bedG <- data.frame(chromosome='chr1', start=1, end=1)
        }
        utils::write.table(bedG, file=savefile.masked.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=TRUE, sep='\t')
        close(savefile.masked.gz)
        stopTimedMessage(ptm)
    }

    ## Write hotspots to file
    if (!is.null(hotspots) & length(hotspots) > 0) {
        hotspots <- insertchr(hotspots)
        savefile.hotspots <- file.path(outputDirectory, paste0(index, '_hotspots.bed.gz'))
        ptm <- startTimedMessage("Writing to file ", savefile.hotspots, " ...")
        savefile.hotspots.gz <- gzfile(savefile.hotspots, 'w')
        head_hotspots <- paste('track name=', index,' description=BedGraph_of_hotspots_',index, '_allChr visibility=1 color=255,0,0', sep="")
        utils::write.table(head_hotspots, file=savefile.hotspots.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=FALSE, sep='\t')
        if (length(hotspots)>0) {
            bedG <- as.data.frame(hotspots)[c('chromosome','start','end')]
        } else {
            bedG <- data.frame(chromosome='chr1', start=1, end=1)
        }
        utils::write.table(bedG, file=savefile.hotspots.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=TRUE, sep='\t')
        close(savefile.hotspots.gz)
        stopTimedMessage(ptm)
    }

    ## Write breakpoint summary to file
    if (!is.null(breaksGraph) & length(breaksGraph) > 0) {
        breaksGraph <- insertchr(breaksGraph)
        savefile.breaksGraph <- file.path(outputDirectory, paste0(index, '_BreaksGraph.bed.gz'))
        ptm <- startTimedMessage("Writing to file ", savefile.breaksGraph, " ...")
        savefile.breaksGraph.gz <- gzfile(savefile.breaksGraph, 'w')
        head_breaks <- paste('track type=bedGraph name=', index,' description=BedGraph_of_breakpoints_',index, '_allChr visibility=dense color=125,35,190', sep="")
        utils::write.table(head_breaks, file=savefile.breaksGraph.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=FALSE, sep='\t')
        if (length(breaksGraph)>0) {
            bedG <- as.data.frame(breaksGraph)[c('chromosome','start','end','hits')]
        } else {
            bedG <- data.frame(chromosome='chr1', start=1, end=1)
        }
        utils::write.table(bedG, file=savefile.breaksGraph.gz, row.names=FALSE, col.names=FALSE, quote=FALSE, append=TRUE, sep='\t')
        close(savefile.breaksGraph.gz)
        stopTimedMessage(ptm)
    }
}